activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb

```diff
@@ -507,6 +507,7 @@ def checkout(checkout_timeout = @checkout_timeout)
       # +conn+: an AbstractAdapter object, which was obtained by earlier by
       # calling #checkout on this pool.
       def checkin(conn)
+        conn.assert_connection_reuseable
```

Comment:
Looks like we are resetting connections, but only when they're reaped from the pool.

The pg adapter properly resets the conn in-place (cc0d54bcc09a8ab834041787df69f6795a468b91) but mysql2 adapter aliases it to a reconnect.